### PR TITLE
feature: support api plugin

### PR DESCRIPTION
**Pull request by zhuangqh**
Signed-off-by: zhuangqh <zhuangqhc@gmail.com>

### Ⅰ. Describe what this PR did
Add http api plugin.

* update handler, enable to add, delete, modify the PouchContainer's HTTP API handler.
 Defined as follow:
 ```
import "github.com/alibaba/pouch/apis/server/types"
// APIPlugin provide the ability to extend PouchContainer HTTP API and change how handler behave.
type APIPlugin interface {
	// The default handler of each API would be passed in while starting HTTP server.
	// UpdateHandler could register extra HTTP API to PouchContainer server,
	// change the behavior of the default handler.
	UpdateHandler([]*types.HandlerSpec) []*types.HandlerSpec
}
```

also refactor metrics, cri api to a unified format
### Ⅱ. Does this pull request fix one issue?

related to #2406

### CODE REVIEW

**Comment by zhuangqh** [comment]
I've refactor them into one method called `UpdateHandler` after discussing with @rudyfly  offline. Thanks for your attention.😄 @xiaoxubeii

File: apis/server/router.go [hunk 1]
```diff
@@ -197,6 +198,8 @@ func filter(handler handler, s *Server) http.HandlerFunc {
 		if req.TLS != nil && len(req.TLS.PeerCertificates) > 0 {
 			issuer := req.TLS.PeerCertificates[0].Issuer.CommonName
 			clientName := req.TLS.PeerCertificates[0].Subject.CommonName
+			ctx = utils.SetTLSIssuer(ctx, issuer)
```
Comment (zhuangqh) [inline apis/server/router.go]: TLS issuer and TLS common name enable API plugin to identify where the HTTP request from

File: apis/server/router.go [hunk 2]
```diff
@@ -8,103 +8,107 @@ import (
...
+	var handlers []*serverTypes.HandlerSpec
+
 	// system
-	s.addRoute(r, http.MethodGet, "/_ping", s.ping)
-	s.addRoute(r, http.MethodGet, "/info", s.info)
-	s.addRoute(r, http.MethodGet, "/version", s.version)
-	s.addRoute(r, http.MethodPost, "/auth", s.auth)
-	s.addRoute(r, http.MethodGet, "/events", withCancelHandler(s.events))
+	handlers = addHandlerSpec(handlers, http.MethodGet, "/_ping", s.ping)
+	handlers = addHandlerSpec(handlers, http.MethodGet, "/info", s.info)
```
Comment (HusterWan) [inline apis/server/router.go]: make me feel not good. 😢 
Can we use more proper policy to add those handlers?

File: apis/server/cri_bridge.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	"context"
+	"net/http"
+)
+
+func (s *Server) criExec(context context.Context, rw http.ResponseWriter, req *http.Request) (err error) {
+	if s.StreamRouter == nil {
+		return EncodeResponse(rw, http.StatusNotImplemented, nil)
+	}
+	s.StreamRouter.ServeExec(rw, req)
```
Comment (HusterWan) [inline apis/server/cri_bridge.go]: `s.StreamRouter.ServeExec(rw, req)`, Why return errors ?

File: docs/features/pouch_with_plugin.md [hunk 4]
```diff
@@ -98,6 +98,23 @@ type CriPlugin interface {
 }
 ```
 
+### api plugin
+
+* update handler, enable to add, delete, modify the PouchContainer's HTTP API handler.
```
Comment (HusterWan) [inline docs/features/pouch_with_plugin.md]: lint error ?

File: apis/server/cri_bridge.go (hunk 3 above)
Comment (HusterWan) [reply apis/server/cri_bridge.go]: I still think a router handler should return an error, WDYT?   @zhuangqh @Starnop 

But I think we no need fix this in this PR

File: pkg/utils/identify.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+const (
+	// PouchTLSIssuer is the key of tls issue stored in context.
+	PouchTLSIssuer TLSKey = "pouch.server.tls.issuer"
+	// PouchTLSCommonName is the key of tls common name stored in context.
+	PouchTLSCommonName TLSKey = "pouch.server.tls.cn"
+)
+
+// SetTLSIssuer set issuer name of tls to context.
+func SetTLSIssuer(ctx context.Context, issuer string) context.Context {
```
Comment (HusterWan) [inline pkg/utils/identify.go]: Just asking, can we add test case for this package?

File: apis/server/system_bridge.go [hunk 6]
```diff
@@ -11,6 +11,7 @@ import (
 	"github.com/alibaba/pouch/apis/types"
 	"github.com/alibaba/pouch/pkg/httputils"
 	"github.com/alibaba/pouch/pkg/utils"
+	util_metrics "github.com/alibaba/pouch/pkg/utils/metrics"
```
Comment (rudyfly) [inline apis/server/system_bridge.go]: Any other named `metrics`?

File: hookplugins/APIPlugin.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
+package hookplugins
+
+import (
+	"github.com/alibaba/pouch/apis/server/types"
+)
+
+// APIPlugin provide the ability to extend PouchContainer HTTP API and change how handler behave.
```
Comment (rudyfly) [inline hookplugins/APIPlugin.go]: s/provide/provides

File: hookplugins/APIPlugin.go [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+type APIPlugin interface {
+	// The default handler of each API would be passed in while starting HTTP server.
```
Comment (rudyfly) [inline hookplugins/APIPlugin.go]: This line should put it below.

**Pull request merged**